### PR TITLE
refactor!: api http statuscode

### PR DESCRIPTION
**Pull request by mabdh**
- Add `mockery` for mocking interface
- For APIs that require Auth Email header, return 403 Forbidden if not present/user does not exist
- Fix error type naming
- Create a function to check empty string
- Add tests in handler

Status code changes:
- Bad request -> 400
- Conflict/Insert existing resource -> 409
- Unauthorized (in check permission) -> 403 (not 401 since that is used when not able to establish identity from credentials or credentials are missing)
- No identity header -> 401
- Server related error -> 500

### CODE REVIEW

**Comment by mabdh** [comment]
Need to hold this until v0.3.x is stable enough

**Comment by mabdh** [comment]
This can be reviewed

File: core/group/service.go
Comment (spy16) [inline core/group/service.go:82]: just a side note on this `str.IsStringEmpty` function. It is advertised and exposed as a public "generic" function. It has no document comment as well. But it does a `TrimSpace()` before doing an empty check. So it's not really generic. And using `!str.IsStringEmtpy()` in place of `grp.ID != ""` is actually changing the behaviour slightly (In this particular case, it is for the better). 

Usually it's better to avoid trying to extract everything into "reusable" functions. Just doing `strings.TrimSpace(val) != ""` is good enough as well.

File: core/group/service.go
Comment (spy16) [inline core/group/service.go:112]: merge this if into the above error handling block ? (all error handling together feels better, for me at-least).

```golang
if err != nil {
   return nil, err
} else if !isAllowed {
   return nil, errors.ErrForbidden
}
```

```

File: core/group/service.go
Comment (spy16) [inline core/group/service.go:113]: why are we returning `[]user.User{}` ? We should be returning `nil` here.

File: core/group/service.go
Comment (spy16) [inline core/group/service.go:198]: this `if` block can be merged to the above error handling block (to keep all error handling together)

File: core/organization/service.go
Comment (spy16) [inline core/organization/service.go:85]: in lot of places we are returning `[]user.User{}`. Wherever there's a need to use empty slice, we should be using `nil`. It prevents an allocation. And has no difference (It cannot cause any nil pointer issues).

⚠️ Only exception to this rule is during JSON marshal. Because `nil` slice value is written out as `null` and empty-slice is marshaled as `[]` (json array)

Comment (mabdh) [reply core/organization/service.go:85]: @spy16 Agree on this, we could just return nil for the empty slices. I keep that part apart from the refactor and this PR just want to don't want to have more-huge changes on this. 

This could be addressed in separate PR and issue.

File: core/group/service.go
Comment (mabdh) [reply core/group/service.go:113]: answered here
https://github.com/odpf/shield/pull/146#discussion_r948686861

File: core/group/service.go
Comment (mabdh) [reply core/group/service.go:82]: I think you have a fair point on that. I think this one is not necessary to be extracted out. Keeping that inline would make this more readable. Thanks for the input @spy16

File: core/group/service.go
Comment (mabdh) [reply core/group/service.go:112]: For this, even I think the `s.relationService.CheckPermission` only needs to return an error. Returning the status `isAlowed` or `isAuthorized` is not really needed. Inside, it could check end return error and we can check the type of error in this CheckPermission or maybe in the handler.
But again, don't want to change the logic that much.

I would actually prefer not to use `else` because sometimes it is confusing but I think we can merge the error handling block together.

File: core/organization/service.go
Comment (spy16) [reply core/organization/service.go:85]: makes sense.

File: internal/api/v1beta1/user.go
Comment (spy16) [inline internal/api/v1beta1/user.go:151]: Do we do any other sanitisation steps? I don't know if this is user-provided value or not.

But usually, it's better to do it like this:
```golang
email = strings.TrimSpace(email)
if email == "" {
   return errors.New("invalid email")
}
```

Otherwise, I can send `"           foo@bar.com         "` and get stored.

Comment (mabdh) [reply internal/api/v1beta1/user.go:151]: @spy16 This is user provided value. No, we mostly don't have any input checking in shield. Previously, for some domains you can even send an empty string as an id and stored in the DB.  I just added several points but maybe not covering all of them.
Agree on that, we better validate the input and return early or do sanitization. There should be another issue that focuses on this. For the email case in this PR, let me refactor as you suggested.

Comment (spy16) [reply internal/api/v1beta1/user.go:151]: yea. may be create an issue about validation so that we can pick it up separately. definitely better not to add this into scope of this PR 👍

**Review by spy16 — APPROVED** [review]
Apart from few validation checks etc (which will be fixed in separate issue), this looks fine to me. But would recommend waiting for at-least 1 other reviewer since there are lot of changes here.

**Pull request merged**
